bindings/rust/s2n-tls/src/fingerprint.rs

```diff
@@ -0,0 +1,497 @@
+    }
+
+    pub fn builder(method: FingerprintType) -> Result<Builder, Error> {
+        Builder::new(method)
+    }
+}
+
+impl Drop for Fingerprint<'_> {
+    /// Resets the underlying [Builder] for the next fingerprint.
+    fn drop(&mut self) {
+        unsafe {
+            s2n_fingerprint_reset(self.0.as_ptr())
+                .into_result()
+                .unwrap()
+        };
+    }
+}
+
+/// Builder that can be reused for multiple fingerprints.
+///
+/// The `Builder` can build a new [Fingerprint] as soon as the old [Fingerprint]
+/// goes out of scope. The [Fingerprint] implementation of [`Self::drop()`] ensures that this
+/// is safe by reseting the underlying C structure.
+///
+/// See the below example:
+/// ```no_run
+/// use s2n_tls::client_hello::ClientHello;
+/// use s2n_tls::fingerprint::{Fingerprint, FingerprintType};
+///
+/// let mut builder = Fingerprint::builder(FingerprintType::JA3)?;
+/// let client_hello_bytes = [ 0, 1, 2, 3, 4, 5 ];
+///     
+/// for _ in 1..10 {
+///     let client_hello = ClientHello::parse_client_hello(&client_hello_bytes)?;
+///     let mut fingerprint = builder.build(&client_hello)?;
+///
+///     let hash_size = fingerprint.hash_size()?;
+///     let mut hash = String::with_capacity(hash_size as usize);
+///     fingerprint.hash(&mut hash)?;
+/// }
+/// # Ok::<(), std::io::Error>(())
+/// ```
+///
+/// Multiple Fingerprints constructed from the same Builder can't exist simultaneously.
+/// The below example fails to compile:
+/// ```compile_fail,E0499
+/// use s2n_tls::client_hello::ClientHello;
+/// use s2n_tls::fingerprint::{Fingerprint, FingerprintType};
+///
+/// let mut builder = Fingerprint::builder(FingerprintType::JA3)?;
+///
+/// let client_hello_bytes = [ 0, 1, 2, 3, 4, 5 ];
+/// let client_hello_1 = ClientHello::parse_client_hello(&client_hello_bytes)?;
+/// let client_hello_2 = ClientHello::parse_client_hello(&client_hello_bytes)?;
+///
+/// let mut fingerprint_1 = builder.build(&client_hello_1)?;
+/// let mut fingerprint_2 = builder.build(&client_hello_2)?;
+///
+/// # Ok::<(), std::io::Error>(())
+/// ```
+pub struct Builder(NonNull<s2n_fingerprint>);
+
+impl Builder {
+    /// Creates a reusable [Builder].
+    pub fn new(method: FingerprintType) -> Result<Self, Error> {
+        crate::init::init();
+        let ptr = unsafe { s2n_fingerprint_new(method.into()).into_result() }?;
+        Ok(Builder(ptr))
+    }
+
+    /// Creates a fingerprint operation for a given [ClientHello].
+    pub fn build(&mut self, client_hello: &ClientHello) -> Result<Fingerprint, Error> {
+        unsafe {
+            s2n_fingerprint_set_client_hello(self.0.as_ptr(), client_hello.deref_mut_ptr())
+                .into_result()
+        }?;
+        Ok(Fingerprint(&mut self.0))
+    }
+}
+
+impl Drop for Builder {
+    /// Frees the memory associated with this [Builder].
+    fn drop(&mut self) {
+        let mut ptr: *mut s2n_fingerprint = unsafe { self.0.as_mut() };
+        unsafe {
```

Comment:
For this drop impl, I think I'd prefer ignoring the failure rather than panicking?

https://github.com/aws/s2n-tls/blob/ed04a0803a93525f4344356819080a6e373b0bc5/bindings/rust/s2n-tls/src/client_hello.rs#L390-L394

Even if the free fails, the program should still be in a well-defined state so it seems safe to continue. I'm not sure that the same can be said for Fingerprint::Drop, so maybe that one should stay as a panic?